pallets/parachain-staking/src/lib.rs

```diff
@@ -0,0 +1,1163 @@
+			T::Currency::reserve(&acc, amount)?;
+			if state.is_active() {
+				Self::update_active(collator.clone(), new_total);
+			}
+
+			// update states
+			let TotalStake {
+				collators: total_collators,
+				delegators: total_delegators,
+			} = <Total<T>>::get();
+			<Total<T>>::put(TotalStake {
+				collators: total_collators,
+				delegators: total_delegators.saturating_add(amount),
+			});
+
+			// TODO: Add update of select_top_candidates
+			<CollatorState<T>>::insert(&collator, state);
+			<DelegatorState<T>>::insert(&acc, Delegator::new(collator.clone(), amount));
+			Self::deposit_event(Event::Delegation(acc, amount, collator, new_total));
+
+			Ok(().into())
+		}
+
+		/// Delegate to another collator candidate which results in updating the
+		/// caller's delegation state.
+		///
+		/// NOTE: Expects the caller to have already delegated.
+		#[pallet::weight(0)]
+		// TODO: Fix unit test panic with transactional feature enabled
+		// #[transactional]
+		pub fn delegate_another_candidate(
+			origin: OriginFor<T>,
+			// TODO: Switch to LookupSource
+			collator: T::AccountId,
+			amount: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let acc = ensure_signed(origin)?;
+			if let Some(mut delegator) = <DelegatorState<T>>::get(&acc) {
+				// delegation after first
+				ensure!(amount >= T::MinDelegation::get(), Error::<T>::DelegationBelowMin);
+				ensure!(
+					(delegator.delegations.len() as u32) < T::MaxCollatorsPerDelegator::get(),
+					Error::<T>::ExceedMaxCollatorsPerDelegator
+				);
+
+				// prepare new collator state
+				let mut state = <CollatorState<T>>::get(&collator).ok_or(Error::<T>::CandidateDNE)?;
+				ensure!(
+					delegator.add_delegation(Bond {
+						owner: collator.clone(),
+						amount
+					}),
+					Error::<T>::AlreadyDelegatedCollator
+				);
+				let delegation = Bond {
+					owner: acc.clone(),
+					amount,
+				};
+				ensure!(state.delegators.insert(delegation.clone()), Error::<T>::DelegatorExists);
+
+				// update state and potentially kick a delegator with less bonded amount
+				state = if (state.delegators.len() as u32) > T::MaxDelegatorsPerCollator::get() {
+					Self::do_update_delegator(delegation, state)?
+				} else {
+					state.total = state.total.saturating_add(amount);
+					state
+				};
+				let new_total = state.total;
+
+				// lock bond
+				// TODO: Switch from reserve to lock
+				T::Currency::reserve(&acc, amount)?;
+				if state.is_active() {
+					Self::update_active(collator.clone(), new_total);
+				}
+
+				// Update states
+				Total::<T>::mutate(|old| {
+					old.delegators = old.delegators.saturating_add(amount);
+				});
+
+				// TODO: Add update of select_top_candidates
+				<CollatorState<T>>::insert(&collator, state);
+				<DelegatorState<T>>::insert(&acc, delegator);
+				Self::deposit_event(Event::Delegation(acc, amount, collator, new_total));
+
+				Ok(().into())
+			} else {
+				Err(Error::<T>::NotYetDelegating.into())
```

Comment:
Maybe move this at the beginning with an `ensure!`?

Reply:
Definitely!
```rust
let mut delegator = <DelegatorState<T>>::get(&acc).ok_or(Error::<T>::NotYetDelegating)?;
```

Reply:
Fixed in af20eb0